### ISSUE TITLE
Poor gesture detection

### ISSUE THREAD

**Issue by Sartharon** [Hikari9/comfortable-swipe#68]
Sometimes gestures don't get triggered after not using one for a while.
Doing the gesture mutiple times works. Every gesture shortly after a working one triggers without any problem.` $ comfortable-swipe debug` shows a gesture (GESTURE_TAP_BEGIN, GESTURE_TAP_UPDATE, GESTURE_TAP_END; GESTURE_SWIPE_BEGIN, GESTURE_SWIPE_UPDATE, GESTURE_SWIPE_END) but nothing triggers (No output in the terminal after `$ comfortable-swipe restart`).

I'm using a Surface Pro 6 with the linux-surface kernel from jakeday (https://github.com/jakeday/linux-surface)
```
$ lsb_release -a 
No LSB modules are available.
Distributor ID:	Deepin
Description:	Deepin 15.9.3
Release:	15.9.3
Codename:	unstable
```

```
$ g++ --version
g++ (Debian 7.3.0-19) 7.3.0
```

```
$ ls -l /dev/input/event*
crw-rw---- 1 root input 13, 64 Apr 24 17:21 /dev/input/event0
crw-rw---- 1 root input 13, 65 Apr 24 17:21 /dev/input/event1
crw-rw---- 1 root input 13, 74 Apr 24 17:21 /dev/input/event10
crw-rw---- 1 root input 13, 75 Apr 24 17:21 /dev/input/event11
crw-rw---- 1 root input 13, 76 Apr 24 17:21 /dev/input/event12
crw-rw---- 1 root input 13, 77 Apr 24 17:21 /dev/input/event13
crw-rw---- 1 root input 13, 78 Apr 24 17:21 /dev/input/event14
crw-rw---- 1 root input 13, 79 Apr 24 17:21 /dev/input/event15
crw-rw---- 1 root input 13, 80 Apr 24 17:21 /dev/input/event16
crw-rw---- 1 root input 13, 81 Apr 24 17:21 /dev/input/event17
crw-rw---- 1 root input 13, 82 Apr 24 17:21 /dev/input/event18
crw-rw---- 1 root input 13, 83 Apr 24 17:21 /dev/input/event19
crw-rw---- 1 root input 13, 66 Apr 24 17:21 /dev/input/event2
crw-rw---- 1 root input 13, 84 Apr 24 17:21 /dev/input/event20
crw-rw---- 1 root input 13, 85 Apr 24 17:21 /dev/input/event21
crw-rw---- 1 root input 13, 86 Apr 24 17:21 /dev/input/event22
crw-rw---- 1 root input 13, 87 Apr 24 17:21 /dev/input/event23
crw-rw---- 1 root input 13, 88 Apr 24 17:21 /dev/input/event24
crw-rw---- 1 root input 13, 89 Apr 24 17:21 /dev/input/event25
crw-rw---- 1 root input 13, 90 Apr 24 17:21 /dev/input/event26
crw-rw---- 1 root input 13, 91 Apr 24 17:21 /dev/input/event27
crw-rw---- 1 root input 13, 92 Apr 24 17:21 /dev/input/event28
crw-rw---- 1 root input 13, 93 Apr 24 17:21 /dev/input/event29
crw-rw---- 1 root input 13, 67 Apr 24 17:21 /dev/input/event3
crw-rw---- 1 root input 13, 68 Apr 24 17:21 /dev/input/event4
crw-rw---- 1 root input 13, 69 Apr 24 17:21 /dev/input/event5
crw-rw---- 1 root input 13, 70 Apr 24 17:21 /dev/input/event6
crw-rw---- 1 root input 13, 71 Apr 24 17:21 /dev/input/event7
crw-rw---- 1 root input 13, 72 Apr 24 17:21 /dev/input/event8
crw-rw---- 1 root input 13, 73 Apr 24 17:21 /dev/input/event9
```

```
$ lsmod | grep hid
hid_sensor_gyro_3d     20480  0
hid_sensor_als         20480  1
hid_sensor_rotation    20480  0
hid_sensor_accel_3d    20480  1
hid_sensor_trigger     20480  9 hid_sensor_gyro_3d,hid_sensor_als,hid_sensor_accel_3d,hid_sensor_rotation
industrialio_triggered_buffer    16384  4 hid_sensor_gyro_3d,hid_sensor_als,hid_sensor_accel_3d,hid_sensor_rotation
hid_sensor_iio_common    20480  5 hid_sensor_gyro_3d,hid_sensor_trigger,hid_sensor_als,hid_sensor_accel_3d,hid_sensor_rotation
industrialio           73728  9 industrialio_triggered_buffer,hid_sensor_gyro_3d,hid_sensor_trigger,kfifo_buf,hid_sensor_als,hid_sensor_accel_3d,hid_sensor_rotation
mac_hid                16384  0
i2c_hid                28672  0
hid_sensor_hub         24576  6 hid_sensor_gyro_3d,hid_sensor_trigger,hid_sensor_iio_common,hid_sensor_als,hid_sensor_accel_3d,hid_sensor_rotation
intel_ishtp_hid        24576  0
hid_multitouch         28672  0
hid_generic            16384  0
usbhid                 53248  0
intel_ishtp            45056  2 intel_ishtp_hid,intel_ish_ipc
hid                   135168  8 i2c_hid,usbhid,hid_multitouch,hid_sensor_hub,intel_ipts,intel_ishtp_hid,hid_generic,surface_acpi
```

```
$ comfortable-swipe status
/usr/local/share/comfortable-swipe/comfortable-swipe.conf
    threshold is OK  (0.0)
        left3 is OK  (super+Right)
        left4 is OFF
       right3 is OK  (super+Left)
       right4 is OFF
          up3 is OFF
          up4 is OFF
        down3 is OFF
        down4 is OFF
autostart is ON
comfortable-swipe program is RUNNING
```

**Comment by Hikari9** [comment]
@Sartharon can you paste a snippet of the output of
```
comfortable-swipe debug | grep GESTURE_SWIPE
```

Preferably from a `GESTURE_SWIPE_BEGIN` all the way to a `GESTURE_SWIPE_END`?

**Comment by Hikari9** [comment]
Your problem might also be related to [jakeday/linux-surface](https://github.com/jakeday/linux-surface/issues/337) kernel itself.

Can you run
```
xinput list
```
then find anything with "Touchpad" then run:
```
xinput list-props "Your Touchpad Name"
```

Then paste your output as a comment? For example, 
```
$ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ Logitech MX Anywhere 2S                 	id=10	[slave  pointer  (2)]
⎜   ↳ Asus TouchPad                           	id=12	[slave  pointer  (2)]   # THIS IS TOUCHPAD
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Power Button                            	id=6	[slave  keyboard (3)]
    ↳ Video Bus                               	id=7	[slave  keyboard (3)]
    ↳ Video Bus                               	id=8	[slave  keyboard (3)]
    ↳ Sleep Button                            	id=9	[slave  keyboard (3)]
    ↳ USB2.0 VGA UVC WebCam: USB2.0 V         	id=11	[slave  keyboard (3)]
    ↳ Asus WMI hotkeys                        	id=13	[slave  keyboard (3)]
    ↳ AT Translated Set 2 keyboard            	id=14	[slave  keyboard (3)]
    ↳ Logitech MX Anywhere 2S                 	id=15	[slave  keyboard (3)]

$ xinput list-props "Asus TouchPad"
    ...
```

**Comment by Sartharon** [comment]
Sorry for the late reply.
I dont think that it's related to the kernel the issue you mentioned concerns the Surface Book 1 which uses a different touchpad. Moreover tapping the touchpad works no problem here.

```
$ xinput list
⎡ Virtual core pointer                    	id=2	[master pointer  (3)]
⎜   ↳ Virtual core XTEST pointer              	id=4	[slave  pointer  (2)]
⎜   ↳ ipts 045E:001F UNKNOWN                  	id=7	[slave  pointer  (2)]
⎜   ↳ ipts 045E:001F Touchscreen              	id=9	[slave  pointer  (2)]
⎜   ↳ ipts 045E:001F Mouse                    	id=10	[slave  pointer  (2)]
⎜   ↳ Microsoft Surface Type Cover Consumer Control	id=13	[slave  pointer  (2)]
⎜   ↳ Microsoft Surface Type Cover Mouse      	id=16	[slave  pointer  (2)]
⎜   ↳ Microsoft Surface Type Cover Touchpad   	id=17	[slave  pointer  (2)]
⎣ Virtual core keyboard                   	id=3	[master keyboard (2)]
    ↳ Virtual core XTEST keyboard             	id=5	[slave  keyboard (3)]
    ↳ Video Bus                               	id=6	[slave  keyboard (3)]
    ↳ ipts 045E:001F                          	id=8	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=11	[slave  keyboard (3)]
    ↳ gpio-keys                               	id=12	[slave  keyboard (3)]
    ↳ Microsoft Surface Type Cover Consumer Control	id=14	[slave  keyboard (3)]
    ↳ Microsoft Surface Type Cover Keyboard   	id=15	[slave  keyboard (3)]
```

```
$ xinput list-props "Microsoft Surface Type Cover Touchpad"
Device 'Microsoft Surface Type Cover Touchpad':
	Device Enabled (153):	1
	Coordinate Transformation Matrix (155):	1.000000, 0.000000, 0.000000, 0.000000, 1.000000, 0.000000, 0.000000, 0.000000, 1.000000
	Device Accel Profile (287):	1
	Device Accel Constant Deceleration (288):	1.000000
	Device Accel Adaptive Deceleration (289):	1.000000
	Device Accel Velocity Scaling (290):	10.000000
	Synaptics Edges (455):	78, 1878, 53, 944
	Synaptics Finger (456):	25, 30, 0
	Synaptics Tap Time (457):	180
	Synaptics Tap Move (458):	96
	Synaptics Tap Durations (459):	180, 180, 100
	Synaptics ClickPad (460):	1
	Synaptics Middle Button Timeout (461):	0
	Synaptics Two-Finger Pressure (462):	282
	Synaptics Two-Finger Width (463):	7
	Synaptics Scrolling Distance (464):	-43, -43
	Synaptics Edge Scrolling (465):	1, 1, 0
	Synaptics Two-Finger Scrolling (466):	1, 0
	Synaptics Move Speed (467):	1.000000, 1.750000, 0.091116, 0.000000
	Synaptics Off (468):	0
	Synaptics Locked Drags (469):	0
	Synaptics Locked Drags Timeout (470):	5000
	Synaptics Tap Action (471):	0, 0, 0, 0, 1, 3, 0
	Synaptics Click Action (472):	1, 3, 2
	Synaptics Circular Scrolling (473):	0
	Synaptics Circular Scrolling Distance (474):	0.100000
	Synaptics Circular Scrolling Trigger (475):	0
	Synaptics Circular Pad (476):	0
	Synaptics Palm Detection (477):	0
	Synaptics Palm Dimensions (478):	6, 50
	Synaptics Coasting Speed (479):	20.000000, 50.000000
	Synaptics Pressure Motion (480):	30, 160
	Synaptics Pressure Motion Factor (481):	1.000000, 1.000000
	Synaptics Grab Event Device (482):	0
	Synaptics Gestures (483):	1
	Synaptics Capabilities (484):	1, 0, 0, 1, 1, 0, 0
	Synaptics Pad Resolution (485):	20, 20
	Synaptics Area (486):	0, 0, 0, 0
	Synaptics Soft Button Areas (487):	978, 0, 817, 0, 0, 0, 0, 0
	Synaptics Noise Cancellation (488):	10, 10
	Device Product ID (280):	1118, 2024
	Device Node (279):	"/dev/input/event6"
```
Out put of a non working swipe:
```
$ comfortable-swipe debug | grep GESTURE_SWIPE
event6   GESTURE_SWIPE_END  +2.92s	3                            #No BEGIN after END?
event6   GESTURE_SWIPE_UPDATE  +7.61s	3 -14.81/ 0.73 (-40.03/ 1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.61s	3 -19.18/ 0.49 (-51.84/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.62s	3 -14.81/ 0.49 (-40.03/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.63s	3 -18.69/ 0.49 (-50.52/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.64s	3 -20.88/ 0.24 (-56.43/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.64s	3 -19.42/ 0.49 (-52.49/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.65s	3 -25.73/ 0.24 (-69.55/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.66s	3 -22.34/ 0.24 (-60.37/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.66s	3 -23.79/ 0.24 (-64.30/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.67s	3 -25.73/ 0.00 (-69.55/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.68s	3 -20.39/-0.24 (-55.12/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.69s	3 -25.98/ 0.73 (-70.21/ 1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.70s	3 -18.94/ 0.00 (-51.18/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.70s	3 -24.04/ 0.24 (-64.96/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.71s	3 -17.24/ 0.49 (-46.59/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.72s	3 -20.64/-0.24 (-55.77/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.72s	3 -18.69/ 0.24 (-50.52/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.73s	3 -18.21/-0.49 (-49.21/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.74s	3 -20.64/-0.49 (-55.77/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.75s	3 -14.57/ 0.49 (-39.37/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.75s	3 -17.48/-1.21 (-47.24/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.76s	3 -13.60/-0.97 (-36.75/-2.62 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.77s	3 -11.90/-1.46 (-32.15/-3.94 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.77s	3 -10.68/-0.73 (-28.87/-1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.78s	3 -9.47/-1.21 (-25.59/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.79s	3 -8.74/-1.70 (-23.62/-4.59 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.80s	3 -8.01/-2.19 (-21.65/-5.91 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.80s	3 -5.58/-1.70 (-15.09/-4.59 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.81s	3 -5.58/-2.43 (-15.09/-6.56 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.82s	3 -4.13/-2.19 (-11.15/-5.91 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.83s	3 -2.43/-1.21 (-6.56/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.83s	3 -3.16/-1.94 (-8.53/-5.25 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.84s	3 -3.40/-1.21 (-9.19/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +7.85s	3 -1.46/-0.49 (-3.94/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_END  +7.87s	3

```
Output of a working swipe:

```
 event6   GESTURE_SWIPE_BEGIN  +8.82s	3
 event6   GESTURE_SWIPE_UPDATE  +8.82s	3 -17.97/ 1.94 (-48.56/ 5.25 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.82s	3 -12.62/ 1.94 (-34.12/ 5.25 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.83s	3 -16.51/ 0.97 (-44.62/ 2.62 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.84s	3 -19.91/ 0.97 (-53.81/ 2.62 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.85s	3 -13.11/ 0.73 (-35.43/ 1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.85s	3 -16.27/ 0.00 (-43.96/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.86s	3 -19.18/ 0.00 (-51.84/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.87s	3 -13.60/ 0.49 (-36.75/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.88s	3 -16.51/-0.49 (-44.62/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.88s	3 -18.69/ 0.49 (-50.52/ 1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.89s	3 -15.30/ 0.24 (-41.34/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.90s	3 -15.78/-0.49 (-42.65/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.90s	3 -19.67/-0.49 (-53.15/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.91s	3 -15.05/ 0.00 (-40.68/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.92s	3 -16.51/-0.24 (-44.62/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.93s	3 -21.61/-0.49 (-58.40/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.93s	3 -16.02/ 0.24 (-43.31/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.94s	3 -21.12/-0.49 (-57.09/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.95s	3 -22.58/ 0.24 (-61.02/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.96s	3 -20.64/-0.24 (-55.77/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.96s	3 -23.55/-0.73 (-63.65/-1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.97s	3 -18.69/ 0.24 (-50.52/ 0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.98s	3 -21.36/-0.49 (-57.74/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.98s	3 -16.75/-1.21 (-45.28/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +8.99s	3 -15.54/-1.46 (-41.99/-3.94 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.00s	3 -15.78/-1.94 (-42.65/-5.25 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.01s	3 -13.84/-0.97 (-37.40/-2.62 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.01s	3 -12.14/-0.49 (-32.81/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.02s	3 -12.14/-0.97 (-32.81/-2.62 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.03s	3 -10.93/-1.21 (-29.53/-3.28 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.04s	3 -9.95/-0.49 (-26.90/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.04s	3 -8.25/-0.49 (-22.31/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.05s	3 -7.04/-0.24 (-19.03/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.06s	3 -5.83/-0.24 (-15.75/-0.66 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.07s	3 -4.86/-0.73 (-13.12/-1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.07s	3 -3.40/-0.73 (-9.19/-1.97 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.08s	3 -3.88/-1.94 (-10.50/-5.25 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.09s	3 -2.43/-1.70 (-6.56/-4.59 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.10s	3 -2.67/-1.46 (-7.22/-3.94 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.10s	3 -1.46/-0.49 (-3.94/-1.31 unaccelerated)
 event6   GESTURE_SWIPE_UPDATE  +9.11s	3 -0.73/ 0.00 (-1.97/ 0.00 unaccelerated)
 event6   GESTURE_SWIPE_END  +9.12s	3

```

It seems that there is no `GESTURE_SWIPE_BEGIN` and therefore nothing happens.
Edit: I did this a few times and can confirm that the `GESTURE_SWIPE_BEGIN` is missing for the non working gestures.

**Comment by Hikari9** [comment]
Based on your output, this seems to be a bug of `libinput` itself. I will move this issue to them. In the meantime, I will push new commit for a workaround you can work with. I will send you instructions to that in a separate Pull Request.

Cheers 🎊

**Comment by Hikari9** [comment]
@Sartharon, please see the patch of the referenced Pull Request #69 and fill out the reviewer checklist. Thanks!

**Comment by Hikari9** [comment]
Issue has been solved by #69. Closing this issue
